Add chunked migration using fetchmany

matthewoliver opened the pull request:
This is a much simpler version of chunked migration. And it taking
advantage of sqlalchemy's fetchmany() funtion.

s-t-e-v-e-n-k commented on psql2mysql/__init__.py:
```diff
@@ -167,13 +173,18 @@ def disable_constraints(self):
             "SET SESSION foreign_key_checks='OFF'"
         )
 
+    @profile
```
I'd prefer if this was able to be turned off.

matthewoliver replied:
I might remove the memory profiling. It was there while I was comparing, it's easy to add. So shouldn't be there by default. I'll remove it in the next version.

s-t-e-v-e-n-k commented on psql2mysql/__init__.py:
```diff
@@ -167,13 +173,18 @@ def disable_constraints(self):
-            table.insert(),
-            rows.fetchall()
-        )
+        if self.chunked and self.chunk_size > 0:
```
You already check it's a positive integer in setup(), so I think just if self.chunked is fine here.

matthewoliver replied:
ahh yes. I wanted 0 to be turned off (migrate the whole table) but in debugging (ie sending none to fetchmany) actually set it to 1, so not what I wanted. Added to > 0 to force it to the else, meant to go back and clean it up.

s-t-e-v-e-n-k commented on psql2mysql/__init__.py:
```diff
@@ -223,20 +243,20 @@ def migrate(self):
                     table.name.startswith("alembic_")):
                 continue
 
+            LOG.info("Migrating table: '%s'" % table.name)
```
Why did you move this block?

matthewoliver replied:
Actually I moved the skipping migrate and alembic tables up, because if were skipping them we shouldn't require the target to actually have the tables sitting there empty.

Also the empty mysql neutron db I used didn't have one of these tables (which I got from another source). so it was also causing me issues.

rhafer replied:
Actually, we only skip those tables because it's assumed that the existing mysql database already has them, because the respecitve "db-manage" tool was already run against it. (So they should be in the very same state as the source db already). I think throwing an error if the source has them but the target doesn't is still a good idea.

s-t-e-v-e-n-k commented on psql2mysql/__init__.py:
```diff
@@ -275,6 +295,19 @@ def setupTypeDecorators(self, srcTable, targetTable):
+            value = int(value)
+
+        if value < 0:
+            raise ValueError()
```
Hmmm. This is almost a little dirty. Surely there is a more pythonic way ...

matthewoliver replied:
Happy to go rework it :) this came up while I was debugging, I should have also written a test at the same time, so bad me :)

s-t-e-v-e-n-k commented on psql2mysql/__init__.py:
```diff
@@ -289,6 +322,15 @@ def add_subcommands(subparsers):
+
+    parser.add_argument("--chunked",
+                        action='store_true',
+                        default=False,
```
Given the MariaDB constraints that we have about writing records, I think this should be on by default.

jsuchome commented on psql2mysql/__init__.py:
```diff
@@ -289,6 +322,15 @@ def add_subcommands(subparsers):
         'migrate',
         help='Migrate data from PostgreSQL to MariaDB')
+
+    parser.add_argument("--chunked",
```
Shouldn't this be added directly into`cli_opts`

matthewoliver replied:
I added it here as these options only relate to the migrate sub command, if I add it at `cli_opts` they'll appear as global options.

jsuchome replied:
OK, I guess you are right

jsuchome commented:
Could you please add a note to README file?

matthewoliver commented:
I've updated, added some tests, updated the doc. And moving the option to cli_opts made some testing better. 
Is there a reason why this was closed.

jsuchome commented:
Sorry for closing, I just hit the wrong button

rhafer commented:
> Who can reopen this? I can't, the "reopen" button is grayed out for me cry

@nicolasbock IIRC we all that that issue. So Matt reopened it as https://github.com/SUSE/psql2mysql/pull/33 (merged meanwhile).

nicolasbock commented:
@rhafer Ah, sorry I missed that. So it's all good then :smile:
